src/common/servers/serverManager.ts

```diff
@@ -0,0 +1,426 @@
+        this.servers.set(newServer.id, newServer);
+        this.serverOrder.push(newServer.id);
+        const tabOrder: string[] = [];
+        getDefaultTabs().forEach((tab) => {
+            const newTab = this.getTabView(newServer, tab.name, tab.isOpen);
+            this.tabs.set(newTab.id, newTab);
+            tabOrder.push(newTab.id);
+        });
+        this.tabOrder.set(newServer.id, tabOrder);
+
+        this.emit(SERVERS_URL_MODIFIED, [newServer.id]);
+        this.persistServers();
+        return newServer;
+    }
+
+    editServer = (serverId: string, server: Team) => {
+        const existingServer = this.servers.get(serverId);
+        if (!existingServer) {
+            return;
+        }
+
+        let urlModified;
+        if (existingServer.url.toString() !== urlUtils.parseURL(server.url)?.toString()) {
+            urlModified = () => this.emit(SERVERS_URL_MODIFIED, [serverId]);
+        }
+        existingServer.name = server.name;
+        existingServer.updateURL(server.url);
+        this.servers.set(serverId, existingServer);
+
+        this.tabOrder.get(serverId)?.forEach((tabId) => {
+            const tab = this.tabs.get(tabId);
+            if (tab) {
+                tab.server = existingServer;
+                this.tabs.set(tabId, tab);
+            }
+        });
+
+        urlModified?.();
+        this.persistServers();
+    }
+
+    removeServer = (serverId: string) => {
+        this.tabOrder.get(serverId)?.forEach((tabId) => this.tabs.delete(tabId));
+        this.tabOrder.delete(serverId);
+        this.lastActiveTab.delete(serverId);
+
+        const index = this.serverOrder.findIndex((id) => id === serverId);
+        this.serverOrder.splice(index, 1);
+        this.remoteInfo.delete(serverId);
+        this.servers.delete(serverId);
+
+        this.persistServers();
+    }
+
+    toggleTab = (tabId: string, isOpen: boolean) => {
+        const tab = this.tabs.get(tabId);
+        if (!tab) {
+            return;
+        }
+        tab.isOpen = isOpen;
+
+        this.persistServers();
+    }
+
+    updateLastActive = (tabId: string) => {
+        const tab = this.tabs.get(tabId);
+        if (!tab) {
+            return;
+        }
+        this.lastActiveTab.set(tab.server.id, tabId);
+
+        this.currentServerId = tab.server.id;
+
+        const serverOrder = this.serverOrder.findIndex((srv) => srv === tab.server.id);
+        this.persistServers(serverOrder);
+    }
+
+    reloadFromConfig = () => {
+        const serverOrder: string[] = [];
+        Config.predefinedTeams.forEach((team) => {
+            const id = this.initServer(team, true);
+            serverOrder.push(id);
+        });
+        if (Config.enableServerManagement) {
+            Config.localTeams.sort((a, b) => a.order - b.order).forEach((team) => {
+                const id = this.initServer(team, false);
+                serverOrder.push(id);
+            });
+        }
+        this.filterOutDuplicateTeams();
```

Comment:
can you share a bit of context as to why would it be possible to have duplicate teams?

Reply:
We have 3 different sources of servers in the app: the build config (servers configured at build time), Group Policy for Windows (for customers mostly) and then the user configuration.

It's possible that a user could add a server that's already pre-configured, or could edit their local config file to do this, causing a duplicate. This can cause some weird behaviour because then the app doesn't know where to send deep links, and the check for which tab to switch to gets fuzzy as well, so we filter out ones that are clear duplicates.